File: companies/Pure Storage/593.valid-square.java
import java.util.*;

/**
 * Algorithm:
 * A valid square has four equal nonzero side distances and two equal diagonal
 * distances, with each diagonal equal to twice a side distance. Compute all six
 * squared distances and sort them.
 *
 * Complexity:
 * Time O(1), space O(1).
 */
class Solution {
    public boolean validSquare(int[] p1, int[] p2, int[] p3, int[] p4) {
        int[][] points = {p1, p2, p3, p4};
        int[] dists = new int[6];
        int idx = 0;
        for (int i = 0; i < 4; i++) {
            for (int j = i + 1; j < 4; j++) {
                dists[idx++] = dist(points[i], points[j]);
            }
        }
        Arrays.sort(dists);
        return dists[0] > 0 &&
               dists[0] == dists[1] &&
               dists[1] == dists[2] &&
               dists[2] == dists[3] &&
               dists[4] == dists[5] &&
               dists[4] == 2 * dists[0];
    }

    private int dist(int[] a, int[] b) {
        int dx = a[0] - b[0];
        int dy = a[1] - b[1];
        return dx * dx + dy * dy;
    }
}

File: companies/Google/593.valid-square.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
    int dist2(const vector<int>& a, const vector<int>& b) {
        int dx = a[0] - b[0], dy = a[1] - b[1];
        return dx * dx + dy * dy;
    }

public:
    bool validSquare(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        vector<vector<int>> p = {p1, p2, p3, p4};
        vector<int> d;
        for (int i = 0; i < 4; ++i) for (int j = i + 1; j < 4; ++j) d.push_back(dist2(p[i], p[j]));
        sort(d.begin(), d.end());
        return d[0] > 0 && d[0] == d[1] && d[1] == d[2] && d[2] == d[3] && d[4] == d[5] && d[4] == 2 * d[0];
    }
};

/*
Interview explanation:
A square has four equal positive side distances and two equal diagonal distances, with diagonal squared equal to twice side squared.

C++ data structures: vector<int> holds the six squared pairwise distances.

Edge cases: duplicate points produce zero side length and are rejected.

Complexity: O(1) time and space.
*/
